### PR TITLE
challenge5

### PR DESCRIPTION
**Pull request by nobeeakon**
updated Routes, added private routes (watch videos, selected videos), added watch later funcionality, added logIn funcionality, improved 404 error page

### CODE REVIEW

**Comment by nobeeakon** [comment]
Acceptance criteria:

  [X]   All the sections have their own route.
 [X]   Navigation across the sections is implemented correctly.
  [X]  The Global State is persistent across all the sections.
  [X]  The "Mocked Authentication" mechanism works correctly.
  [X]  The "session data" is stored in the Global Context correctly.
  [X]  Videos can be added to the Favorites list. (_Implemented a watch later option, instead_)
   [X] Videos can be removed from the Favorites list.
   [X] Navigation to Favorite Videos View using a private route is implemented correctly (only authenticated users should access this section).
   [X] Navigation to the Favorite Video Details View using a private route is implemented correctly (only authenticated users should access this section).
   [X] Information for the selected favorite video is displayed correctly
   [X] The list of other favorite videos in the Favorite Video Details View is displayed.
   [X]  Testing coverage is above 70%. (Please include a screenshot of the code coverage report in your PR description). (_Almost there :)_ )

Bonus Points

   [X]  The Add/remove from favorites button should appear when the user passes the mouse over the video card in the list.
   [ ] Integrate with a real authentication provider (such as Auth0, OAuth or Firebase).
   [X] A 404 section is shown when a route is not found.
   [X]  The Login View is implemented as a modal using React Portals.

image

**Comment by nobeeakon** [comment]
Acceptance criteria:

[X] All the sections have their own route.
[X] Navigation across the sections is implemented correctly.
[X] The Global State is persistent across all the sections.
[X] The "Mocked Authentication" mechanism works correctly.
[X] The "session data" is stored in the Global Context correctly.
[X] Videos can be added to the Favorites list. (Implemented a watch later option, instead)
[X] Videos can be removed from the Favorites list.
[X] Navigation to Favorite Videos View using a private route is implemented correctly (only authenticated users should access this section).
[X] Navigation to the Favorite Video Details View using a private route is implemented correctly (only authenticated users should access this section).
[X] Information for the selected favorite video is displayed correctly
[X] The list of other favorite videos in the Favorite Video Details View is displayed.
[X] Testing coverage is above 70%. (Please include a screenshot of the code coverage report in your PR description). 

coverage_last


Bonus Points

[X] The Add/remove from favorites button should appear when the user passes the mouse over the video card in the list.
[ X] Integrate with a real authentication provider (such as Auth0, OAuth or Firebase).
[X] A 404 section is shown when a route is not found.
[X] The Login View is implemented as a modal using React Portals.

File: src/components/NavBar/LoginSignIn/LogInComponent/Login.component.jsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  };
+
+  if (isAuthenticated) closeModal();
+
+  const handleUserPassword = (e) => {
+    e.preventDefault();
+    e.stopPropagation();
+
+    const tmpPassword = e.target.value;
+    setUserPassword(tmpPassword);
```
Comment (vmcruz) [inline src/components/NavBar/LoginSignIn/LogInComponent/Login.component.jsx:38]: You can avoid this variable if you pass the value directly to the `setUserPassword`.

File: src/components/NavBar/LoginSignIn/LogInComponent/Login.component.jsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  const handleSubmit = async (event) => {
+    event.preventDefault();
+    event.stopPropagation();
+
+    if (!isEmailAndPasswordFilled) {
+      setErrorOnSubmit(true);
+    } else {
+      try {
+        const userData = await auth.signInWithEmailAndPassword(email, userPassword);
```
Comment (vmcruz) [inline src/components/NavBar/LoginSignIn/LogInComponent/Login.component.jsx:51]: Great. I see you implemented your own login mechanism for the project. I tested it over the netlify app and it worked properly.

File: src/components/NavBar/LoginSignIn/LogInComponent/Login.component.jsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            <Styled.UserPasswordInput
+              id="user-password"
+              onChange={handleUserPassword}
+              value={userPassword}
+            />
+          </Styled.InputField>
+
+          <Styled.ButtonContainer>
+            <Styled.InputButton disabled={!isEmailAndPasswordFilled}>
+              Log-in
```
Comment (vmcruz) [inline src/components/NavBar/LoginSignIn/LogInComponent/Login.component.jsx:96]: I noticed that the SignInComponent and this LogInComponent shared a lot of the render logic. Maybe consider having a third shared component to reuse in both places, to keep your components as DRYer as possible.

File: src/components/NavBar/LoginSignIn/LogInComponent/Login.spec.jsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+const LoginProviders = ({ children }) => {
+  return (
+    <GlobalContextProvider>
+      <ThemesProvider>{children}</ThemesProvider>
+    </GlobalContextProvider>
+  );
+};
+
+const customRenderLogin = (ui, options) =>
+  render(ui, { wrapper: LoginProviders, ...options });
```
Comment (vmcruz) [inline src/components/NavBar/LoginSignIn/LogInComponent/Login.spec.jsx:22]: Same here with the `customRender`, perhaps consider moving this to an util file, to shared it across tests.

File: src/components/NavBar/LoginSignIn/LogInModal/LoginModal.jsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import Login from '../LogInComponent';
+import SignIn from '../SignInComponent';
+
+import * as Styled from './LoginModal.styled';
+
+const LoginModal = ({ closeModal }) => {
+  const [isNewUser, setIsNewUser] = useState(false);
+  const [buttonMessage, setButtonMessage] = useState('');
+
+  const ModalContent = isNewUser ? SignIn : Login;
```
Comment (vmcruz) [inline src/components/NavBar/LoginSignIn/LogInModal/LoginModal.jsx:14]: With this ternary you know which content will be displayed for your modal, but I find it odd that you needed to pass the `setButtonMessage` to it to set a value on this component. Maybe that message should be part of your modal as well, and pass the message by prop, then render the message inside the `ModalContent`.

File: src/components/NavBar/LoginSignIn/SignInComponent/SignIn.component.jsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            />
+          </Styled.InputField>
+
+          {password.length <= 6 && (
+            <Styled.AlertFlag>Password must have at least 7 characters</Styled.AlertFlag>
+          )}
+
+          {!arePassWordsEqual &&
+            password.length > 6 &&
+            confirmatoryPassword.length > 6 && (
```
Comment (vmcruz) [inline src/components/NavBar/LoginSignIn/SignInComponent/SignIn.component.jsx:136]: These validations were implied in `arePasswordsEqual`. There, you're already checking for the password length. Also, I don't think you should be checking length there, since you have a message specific for that error in line 131. The only concern for this message should be if the passwords are equal or not, regardless the length.

File: src/components/NavBar/NavBarSignInOrAvatar.component.jsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+import Avatar from './SingInAndAvatar/Avatar.component';
+import SignIn from './SingInAndAvatar/SignIn.component';
+
+const SignInOrAvatar = () => {
+  const { globalState } = useGlobalContext();
+
+  const { isAuthenticated } = globalState;
+
+  return <>{isAuthenticated ? <Avatar /> : <SignIn />}</>;
```
Comment (vmcruz) [inline src/components/NavBar/NavBarSignInOrAvatar.component.jsx:13]: This ternary could be moved to the NavBar component where you needed it, you already have the `globalState` over there. It will simplified the code and make it more readable, as you don't need to look for this file just to see the conditional rendering.

File: src/components/NavBar/Private/PrivateModal/Private.modal.jsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+import * as Styled from './Private.modal.styled';
+
+const PrivateModal = (props, ref) => {
+  const { isModalOpen } = props;
+
+  if (!isModalOpen) return null;
+
+  return ReactDom.createPortal(
+    <>
```
Comment (vmcruz) [inline src/components/NavBar/Private/PrivateModal/Private.modal.jsx:14]: You don't need the Fragments in this case, as you have a root component (`Styled.ModalContainer`)

File: src/components/VideoDetail/VideoDetailView/Stats.component.jsx [hunk 9]
```diff
@@ -5,7 +5,9 @@ import { BiDislike, BiLike } from 'react-icons/bi';
 
 import { StatsContainer, StatElement } from './VideoDetailView.styled';
 
-const Stats = ({ views, likes, dislikes }) => {
+import WatchLaterButton from '../../WatchLater/WatchLaterButton';
+
+const Stats = ({ views, likes, dislikes, videoInfo, videoId }) => {
```
Comment (vmcruz) [inline src/components/VideoDetail/VideoDetailView/Stats.component.jsx:10]: Some props were added to the component, but not to the prop types checking.

File: src/components/VideoDetail/VideoDetailView/VideoDetailView.component.jsx [hunk 10]
```diff
@@ -24,23 +23,29 @@ const VideoDetailView = ({
   views,
   likes,
   dislikes,
-  tags,
+  videoInfo,
```
Comment (vmcruz) [inline src/components/VideoDetail/VideoDetailView/VideoDetailView.component.jsx:26]: Same in here

File: src/components/VideoList/VideoList/VideoList.presenter.jsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+
+import * as Styled from './VideoList.styled';
+import VideoListCard from '../VideoListCard';
+
+import getVideoKey from '../../../utils/functions/getVideoKey';
+
+const CardListPresenter = ({ videoList, isInPrivateRoute, removeItem }) => {
```
Comment (vmcruz) [inline src/components/VideoList/VideoList/VideoList.presenter.jsx:8]: Missing `PropTypes` for this component.

File: src/components/VideoList/VideoListCard/VideoListCard.component.jsx [hunk 12]
```diff
@@ -1,74 +1,94 @@
...
+        </Styled.ExtraInfoDiv>
+      </Styled.InfoContainer>
+    </Styled.VideoCardContainer>
   );
 };
 
 VideoListCard.propTypes = {
   title: PropTypes.string,
-  author: PropTypes.string,
+  channelTitle: PropTypes.string,
```
Comment (vmcruz) [inline src/components/VideoList/VideoListCard/VideoListCard.component.jsx:83]: Some other props are missing for this component as well.

File: src/components/VideoList/VideoListCard/VideoListCard.component.jsx [hunk 13]
```diff
@@ -1,74 +1,94 @@
...
 
   const titleFull = he.decode(title);
-  const authorFull = he.decode(author);
+  const channelTitleFull = he.decode(channelTitle);
   const descriptionFull = he.decode(description);
 
   const shortTitle = titleFull.length > 55 ? `${titleFull.slice(0, 52)}...` : titleFull;
-  const shortAuthor =
-    authorFull.length > 55 ? `${authorFull.slice(0, 52)}...` : authorFull;
+  const shortchannelTitle =
```
Comment (vmcruz) [inline src/components/VideoList/VideoListCard/VideoListCard.component.jsx:41]: Truncating a string is looking more recurrent here. Perhaps consider adding a util that receives a string and a desired length, and it returns the truncated versions if larger, or the original if shorter.

File: src/components/WatchLater/WatchLaterButton/WatchLaterButton.jsx [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    <Styled.StyledWatchLaterButton
+      onClick={handleWatchLater}
+      data-testid="watch-later-button"
+    >
+      {isVideoSaved ? (
+        <>
+          <ImCheckboxChecked data-testid="checked-icon-testid" /> Watch Later
+        </>
+      ) : (
+        <>Watch Later</>
```
Comment (vmcruz) [inline src/components/WatchLater/WatchLaterButton/WatchLaterButton.jsx:58]: You shouldn't need `React.Fragment` here. Take a look by removing it and just using plain strings `'Watch Later'`.

File: src/pages/Watch/Watch.page.presenter.jsx [hunk 15]
```diff
@@ -0,0 +1,57 @@
+import React from 'react';
+
+import { Redirect } from 'react-router-dom';
+
+import VideoDetailView from '../../components/VideoDetail/VideoDetailView';
+
+import * as Styled from './Watch.page.styled';
+
+const WatchPagePresenter = ({ videoId, videoList, isLoading, error, RelatedVideos }) => {
```
Comment (vmcruz) [inline src/pages/Watch/Watch.page.presenter.jsx:9]: Missing PropTypes validations

File: src/pages/Watch/Watch.page.presenter.jsx [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+      <Styled.VideoContainer>
+        <VideoDetailView
+          videoId={videoInfo.videoId}
+          channelTitle={videoInfo.channelTitle}
+          title={videoInfo.title}
+          description={videoInfo.description}
+          views={videoInfo.views}
+          likes={videoInfo.likes}
+          dislikes={videoInfo.dislikes}
+          videoInfo={videoItem}
```
Comment (vmcruz) [inline src/pages/Watch/Watch.page.presenter.jsx:49]: Since you are mapping your props to match the `videoInfo` object you created, you could spread the props like so:

```javascript
<VideoDetailView {...videoInfo} videoInfo={videoItem} />
```

what that does is to spread your object and set your props based on the keyPairs of that object.

File: src/utils/authentication/firebase.js [hunk 17]
```diff
@@ -0,0 +1,16 @@
+import firebase from 'firebase';
+
+const API_SECRET_KEY = process.env.REACT_APP_FB_API_KEY;
```
Comment (vmcruz) [inline src/utils/authentication/firebase.js:3]: Even though the custom login and registration works, remember that react is a client side application, which means that for it to connect to your firebase instance it needs to have access to that key. What I'm trying to say is that don't rely only on having your secrets in the `.env` file, but implement security also on the server side (firebase in this case), and only allow certain domains so that you restrict its usage. Haven't tested if this was your case, but just a FYI for future implementations.

File: src/utils/hooks/useIsMountedRef.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+  useEffect(() => {
+    isMountedRef.current = true;
+
+    return () => {
+      isMountedRef.current = false;
+    };
+  }, []);
+
+  return isMountedRef;
```
Comment (vmcruz) [inline src/utils/hooks/useIsMountedRef.js:15]: This will only handles a boolean value and nothing more inside the `current`. Perhaps you could return `isMountedRef.current` directly. That will give you the direct access to the boolean instead of doing that every time you want to use the hook value.

Comment (vmcruz) [reply src/utils/hooks/useIsMountedRef.js:15]: If you do this, consider renaming the hook to `useIsMounted`.

**Review by vmcruz — COMMENTED** [review]
Left my final comments. Great work!. I like the way you go with the unit tests, concise and you test a lot of edge cases. Awesome work.
